pkg/cloud/metadata/ecs.go

```diff
@@ -30,6 +32,77 @@ type ECSMetadata struct {
 	idDoc InstanceIdentityDocument
 }
 
+type HttpStatusError struct {
+	StatusCode int
+}
+
+func (e HttpStatusError) Error() string {
+	return fmt.Sprintf("unexpected status error: %d", e.StatusCode)
+}
+
+func (e HttpStatusError) Retryable() bool {
+	return e.StatusCode >= 500 || e.StatusCode == http.StatusTooManyRequests ||
+		e.StatusCode == http.StatusUnauthorized // re-fetch token on 401
+}
+
+var ErrInvalidIdentityDoc = errors.New("invalid ECS instance identity document")
+
+type ecsFetchContext struct {
+	token    string
+	tokenReq *http.Request
+	req      *http.Request
+}
+
+func (ctx *ecsFetchContext) fetchToken(client *http.Client) error {
+	resp, err := client.Do(ctx.tokenReq)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return HttpStatusError{resp.StatusCode}
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	ctx.token = string(body)
+	return nil
+}
+
+func (ctx *ecsFetchContext) fetch(client *http.Client) (InstanceIdentityDocument, error) {
+	var doc InstanceIdentityDocument
+	if ctx.token == "" {
```

Comment:
What will happen here if security enhance mode is disabled?

Reply:
ECS metadata server has 3 modes:
1. normal
1. security enhanced only
1. disabled

The current version of CSI only supports normal mode. With this PR, it can support modes 1 & 2. When the user disables metadata server, we can still fallback to other metadata providers.